Common/src/main/groovy/io/github/groovymc/cgl/api/datagen/recipe/GRecipeProvider.groovy

```diff
@@ -0,0 +1,257 @@
+        recipe(new GShapedRecipeBuilder(), clos)
+    }
+
+    protected GShapelessRecipeBuilder shapeless(@DelegatesTo(value = GShapelessRecipeBuilder,strategy = Closure.DELEGATE_FIRST) @ClosureParams(value = SimpleType, options = 'io.github.groovymc.cgl.api.datagen.GShapelessRecipeBuilder') Closure clos) {
+        recipe(new GShapelessRecipeBuilder(), clos)
+    }
+
+    protected <T extends SaveableRecipe> T recipe(@DelegatesTo.Target('recipe') Class<T> recipeClass, @DelegatesTo(target = 'recipe', genericTypeIndex = 0, strategy = Closure.DELEGATE_FIRST) @ClosureParams(FirstParam.FirstGenericType) Closure clos) {
+        return recipe(recipeClass.getDeclaredConstructor().newInstance(), clos)
+    }
+
+    protected <T extends SaveableRecipe> T recipe(@DelegatesTo.Target('recipe') T recipe, @DelegatesTo(target = 'recipe', strategy = Closure.DELEGATE_FIRST) @ClosureParams(FirstParam) Closure clos) {
+        recipe.provider = this
+        clos.resolveStrategy = Closure.DELEGATE_FIRST
+        clos.delegate = recipe
+        clos(recipe)
+        forgottenRecipes.add(recipe)
+        return recipe
+    }
+
+    protected <T extends RecipeBuilder> SaveableRecipe normal(@DelegatesTo.Target('recipe') T recipe, @DelegatesTo(target = 'recipe', strategy = Closure.DELEGATE_FIRST) @ClosureParams(FirstParam) Closure clos = {}) {
+        clos.resolveStrategy = Closure.DELEGATE_FIRST
+        clos.delegate = recipe
+        clos(recipe)
+        return new SaveableRecipe() {
+            @Override
+            void save(ResourceLocation location) {
+                forgottenRecipes.remove(this)
+                recipe.save(getProvider().writer, location)
+            }
+
+            @Override
+            GRecipeProvider getProvider() {
+                return GRecipeProvider.this
+            }
+
+            @Override
+            void setProvider(GRecipeProvider provider) {
+
+            }
+        }
+    }
+}
+
+@CompileStatic
+trait SaveableRecipe {
+    GRecipeProvider provider
+
+    /**
+     * Saves this recipe to the given {@code location}, using the provider's {@link GRecipeProvider#defaultNamespace  default namespace}.
+     */
+    void save(String location) {
+        this.save(location.contains(':') ? new ResourceLocation(location) : new ResourceLocation(provider.defaultNamespace, location))
+    }
+
+    /**
+     * Saves this recipe to the given {@code location}.
+     */
+    abstract void save(ResourceLocation location)
+}
+
+@CompileStatic
+trait BaseRecipeBuilder extends SaveableRecipe implements RecipeBuilder {
+    ItemStack result
+
+    ItemStack result(ItemLike result) {
+        this.result = result.asItem().getDefaultInstance()
+        return this.result
+    }
+
+    ItemStack result(ItemStack result) {
+        this.result = result
+        return this.result
+    }
+
+    Item getResult() {
+        return this.@result.item
+    }
```

Comment:
Could `setResult()` methods be added to allow setter-style DSL similar to some of the other methods?